- CountRabbit: A rapid k-mer-based counting tool for analysis of sequencing data
- What goes in: 
	- The names of your reference file 
	- The location of sequencing datafiles
	- <Optional> choices of parameters for 
- What comes out: 
	-A table of match counts (K-mers and/or Reads) 
	-A very quick visualization
- Important Features: Quick, Simple, Consistent handling of sequences shared between references 
- CountRabbit is K-mer based
	- Any read that matches at least one k-mer in a reference genome is assigned to that genome
	- If only one reference genome is matched, a k-mer or read is considered to be "uniquely" mapped
	- Reads (or K-mers) that match multiple references are considered to be "public"
	- A third (optional) category "local repeat" describes K-mers 
		- present multiple times
	        - but all in one reference sequence
	- With the default K-mer length of 31, specificity of matches is such that false positive will be very rare

- CountRabbit is primarily a tool for counting matches (not for many other tasks):
	- CountRabbit is not designed to 
		- provide strandedness data
		- detect distant homology, or 
		- assemble sequencing data into contigs.
	- CountRabbit is not designed for (or helpful in) detecting mutations, rearrangements, or termini
	- CountRabbit provides a provisional (and customizable) graphic output
		- But this graphic is intended for quick experimental examination
				- publication may require additional tools.
	- Features of CountRabbit overlap with
		- Many excellent and free public tools
		- Several local tools: REVA,HQAlign,Polybench, Jazz18Heap
	- If everything is working properly, CountRabbit can be much faster to implement and/or run
	- CountRabbit is a work in progress
		- For now, any results (or lack thereof) should be checked/verified with other tools
	- Please report any bugs/crashes to afire@stanford.edu
 - For more detail, see the CountRabbitDocumentationFile (CountRabbit_Docs_ba03_072424.docx) in this repository
